xoa/data_source_vms.go

```diff
@@ -0,0 +1,107 @@
+package xoa
+
+import (
+	"log"
+	"strings"
+
+	"github.com/ddelnano/terraform-provider-xenorchestra/client"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func dataSourceXoaVms() *schema.Resource {
+
+	return &schema.Resource{
+		Read: dataSourceVmsRead,
+		Schema: map[string]*schema.Schema{
+			"vms": &schema.Schema{
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     resourceVm(),
+			},
+			"num": &schema.Schema{
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"pool_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"container": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"power_state": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func dataSourceVmsRead(d *schema.ResourceData, m interface{}) error {
+	c := m.(client.XOClient)
+	searchVm := client.Vm{
+		PowerState: d.Get("power_state").(string),
+		Container:  d.Get("container").(string),
+		PoolId:     d.Get("pool_id").(string),
+	}
+
+	vms, err := c.GetVms(searchVm)
+	if err != nil {
+		return err
+	}
+
+	if err = d.Set("vms", vmToMapList(vms)); err != nil {
+		return err
+	}
+	if err = d.Set("num", len(vms)); err != nil {
+		return err
+	}
+	if searchVm.Container != "" {
+		d.SetId(searchVm.Container)
+		return nil
+	}
+	d.SetId(searchVm.PoolId)
```

Comment:
I think its easier to reason about a simpler id mechanism (regarding conditionally using pool or host id). I would prefer if we used a scheme that was flexible enough for both cases.

I think using the pool id will be problematic. Imagine if a user wanted to group a pool's VMs two different ways, the data sources will clash.

I need some time to think about this, but I'll post my ideas once I've had a chance to think more.

Reply:
Why don't we concatenate all the inputs and hash it as the ID?

Something like this
```
powerState := d.Get("power_state").(string)
poolId := d.Get("host").(string)
host := d.Get("host").(string)

[...]

hash := internal.String(fmt.Sprintf(
"%s-%s-%s", powerState, poolId, host
))
# This "internal" package is poorly named but it is the previous hashcode.String function that was deprecated https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk@v1.17.2/helper/hashcode#String
d.SetId(internal.String(fmt.Sprintf(")
```

This will allow users to specify multiple queries as long as it isn't identical and has the benefit of using a consistent ID no matter the types of inputs.

Reply:
Regarding https://github.com/terra-farm/terraform-provider-xenorchestra/pull/161#discussion_r667862367, I'm worried that the ordering of the VMs will change and that will cause the terraform state to constantly update it's ID. By hashing all the inputs that go into the VM filtering we are ensuring that it is deterministic.